### ISSUE TITLE
[Enhancement] Show pending Incoming Xfer 

### ISSUE THREAD

**Issue by p2poolmining** [turtlecoin/turtlecoin-mobile-wallet#85]
Since TonChan is a mobile wallet, that will be used with xfer's between mobile wallet owners, it would be great if an **incoming/outgoing** transfer can be shown in the  LOCKED AMOUNT and included in the balance calculation even before the xfer is completed.

For instance, **zedwallet** will show an incoming locked amount which is added to the wallet balance even though the transaction is still setting in the mempool. This makes the whole transfer process almost instantaneous even though the transaction has yet to be added to a block.

If this same feature could be added to TonChan then this whole process can speed up the ability of the mobile wallet transaction process and provide  a quicker **_"visual confirmation"_** between both the sender and the receiver during a purchase/sale using TonChan.

**Comment by 2acoin** [comment]
Great timing... We need this similar functionality too. Looking at the code to determine the best approach for implementation at the moment.

**Comment by brandonlehmann** [comment]
The problem with showing transactions in the mempool is that:

1) Just because it's in the mempool does not mean it will be included in a block. Mempool transactions are in an *uncommitted* state.
2) When the mempool fills to the brim (say someone optimizes a wallet), the app will be stuck verifying whether any of those transactions are destined for the wallet at hand when chances are the large majority will not. Once they are included in blocks, they *may* be scanned again (and really should be).
3) While the c++ daemon supports returning mempool transaction details, the blockchain cache does not and it is unlikely such support will be added (due to #1 & #2)
4) Transactions in the mempool are akin to "the check is in the mail".
5) With 30s block times, the large majority of transactions are included in blocks within 2 blocks (~60s) -- which is not that much time to wait.
6) etc etc etc

**Comment by ExtraHash** [comment]
what's the reason for this behavior existing in zedwallet if 1 and 2 are such an issue?

although on a mobile processor i could see this being significantly more taxing, i as well wanted to implement this into a desktop wallet

**Comment by brandonlehmann** [comment]
> what's the reason for this behavior existing in zedwallet if 1 and 2 are such an issue?
> 
> although on a mobile processor i could see this being significantly more taxing, i as well wanted to implement this into a desktop wallet

Because zedwallet was a rewrite of simple wallet which does support it. zedwallet++ is also a rewrite of zedwallet which still supports it -- but not when it's talking to a blockchain cache. Honestly, I'm not opposed to lobbying for the removal of it from zedwallet++ due to #2 (for sure) and #1 (a little).

**Comment by zpalmtree** [comment]
zedwallet++ does not support printing transactions in the memory pool.

**Comment by zpalmtree** [comment]
Btw, you need an absolute link when you're linking to another repos issues :wink:

**Comment by brandonlehmann** [comment]
> zedwallet++ does not support printing transactions in the memory pool.

I stand corrected.

**Comment by brandonlehmann** [comment]
> Btw, you need an absolute link when you're linking to another repos issues 😉

I was referring to my list, not repo issues lol.

**Comment by JJD1000** [comment]
This functionality would be great if possible /achievable! 😁

**Comment by 2acoin** [comment]
Yes, let's resurrect this enhancement request. For turlecoin (with a 30 second block time) it is not a big issue. But for forks with greater block times it is beneficial to see that a trxfer is pending...

**Comment by 2acoin** [comment]
We'd be willing to add a PR to the blockchain cache (# 3 in the [list here](https://github.com/turtlecoin/turtlecoin-mobile-wallet/issues/85#issuecomment-515714860) )
so that we can cache all the pending trxs from the mempool.

If that is done, what would it take to include this in zedwallet++ ?

**Comment by ExtraHash** [comment]
The problem is that you need to scan each tx in the mempool to see if it is yours or not. Most phones can hardly keep up with syncing the blocks, this would effectively require you to scan twice as many tx (once when in mempool, once when included in a block)

**Comment by brandonlehmann** [comment]
> We'd be willing to add a PR to the blockchain cache (# 3 in the [list here](https://github.com/turtlecoin/turtlecoin-mobile-wallet/issues/85#issuecomment-515714860) )
> so that we can cache all the pending trxs from the mempool.

Not to derail the conversation; however, I can assure you that I have no plans on supporting or accepting a PR to add mempool transaction information to the cache for the reasons I stated above. If you wish to maintain and operate a fork of the code with those changes, you are of course, free to do so, but it will not make it into the main repo (unless there is a a solid need for it in the TurtleCoin® implementation. The focus of the cache is TurtleCoin®, not forks).

**Comment by 0des** [comment]
@2acoin Those forks should fund a bounty or submit a pull request so that TRTL can continue to provide you with functioning software to use.

**Comment by 2acoin** [comment]
Okay, understand... I believe this is something that we will want to add into 2ACoin (fork of turtlecoin) due to our 90 seconds block time. We have it working in a test instance but have not brought it into our forked version of TurtlePay Blockcache API [2ACoin Blockcache API](https://api.2acoin.org/transaction/pool).

If we were to ask the community for assistance with scanning this set of mempool transaction in our TonChan fork to determine pending payments. Where would be the best place for us to request assistance?

**Comment by brandonlehmann** [comment]
> If we were to ask the community for assistance with scanning this set of mempool transactions in our TonChan fork to determine pending payments. Where would be the best place for us to request assistance?

Does 2aCoin have a discord, chat, forum, or otherwise? That's probably a good place to start.
